[AutoDiff] Revamp differentiation transform.

dan-zheng opened the pull request:
- Directly generate primal code in VJP functions.
  - Rename `PrimalGenCloner` to `VJPEmitter`.
  - This unblocks control-flow support.
    Primal data structures can be generated based on activity analysis.
- Refactor differentiation transform to use iterative `autodiff_function`
  instruction worklist instead of a fixed `DifferentiationTask` worklist.
  - `VJPEmitter::visitApply` now creates `autodiff_function` and
    `autodiff_function_extract` instructions instead of directly emitting
    associated function references.
  - An iterative loop canonicalizes `autodiff_function` instructions,
    promoting them to `@differentiable` function-typed values.
- Add `autodiff_function_extract` folding optimization.
  - Fold `autodiff_function_extract` users of `autodiff_function` instructions,
    directly replacing them with operands of the `autodiff_function`
    instruction.
  - If the `autodiff_function` instruction has only `autodiff_function_extract`
    users, delete the instruction itself after folding.
- Remove unnecessary auxiliary data structures.
  - `DifferentiationTask`: replace with map from `[differentiable]` attributes
    to `DifferentiationInvoker`.
  - `PrimalGen`, `AdjointGen`: replace with `autodiff_function` iterative loop.

dan-zheng committed:
Minor edits.

dan-zheng committed:
[AutoDiff] Revamp differentiation transform.

- Directly generate primal code in VJP functions.
  - Rename `PrimalGenCloner` to `VJPEmitter`.
  - This unblocks control-flow support.
    Primal data structures can be determined based on activity analysis.
- Refactor differentiation transform to use iterative `autodiff_function`
  instruction worklist instead of a fixed `DifferentiationTask` worklist.
  - `VJPEmitter::visitApply` now creates `autodiff_function` and
    `autodiff_function_extract` instructions instead of directly emitting
    associated function references.
  - An iterative loop canonicalizes `autodiff_function` instructions, filling
    in missing associated functions.
- Remove unnecessary auxiliary data structures.
  - `DifferentiationTask`: replace with map from original function and
    `[differentiable]` attribute to `DifferentiationInvoker`.
  - `PrimalGen`, `AdjointGen`: replace with `autodiff_function` iterative loop.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -525,6 +525,15 @@ bool ReabstractionInfo::prepareAndCheck(ApplySite Apply, SILFunction *Callee,
   }
 
+  // SWIFT_ENABLE_TENSORFLOW
+  // Disable specialization for instructions that are operands of
```
This change is necessary to prevent post-differentiation SIL passes from generating invalid SIL.

---

In particular, this change disables specialization of operands to `autodiff_function`.
It may be possible to teach the specializer how to handle `autodiff_function` instructions. I don't know much work that would take, it seems difficult.

dan-zheng commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -74,7 +74,15 @@ static bool foldInverseReabstractionThunks(PartialApplyInst *PAI,
   // partial_apply without any substitutions or arguments is just a
   // thin_to_thick_function.
-  if (!PAI->hasSubstitutions() && (PAI->getNumArguments() == 0)) {
+  if (!PAI->hasSubstitutions() && (PAI->getNumArguments() == 0) &&
```
This change is necessary to prevent post-differentiation SIL passes from generating invalid SIL.

---

This change is rather innocuous and can be upstreamed to `master`. It just adds a previously unexercised check.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2010,61 +2010,133 @@ emitAssociatedFunctionReference(ADContext &context, SILBuilder &builder,
+
 }
 
+enum class DifferentiationThunkKind {
```
Todo: add documentation.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1845,34 +1800,61 @@ emitAssociatedFunctionReference(ADContext &context, SILBuilder &builder,
+          DifferentiationInvoker::Kind::IndirectDifferentiation)
+        contextualRequirements = std::get<2>(
+            invoker.getIndirectDifferentiation())->getRequirements();
+      // Parent function containing instruction
```
Todo: remove stray comment.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2980,26 +3016,97 @@ class PrimalGenCloner final
+      // redesign: currently, each original function + attribute pair is mapped
+      // only to one invoker.
+      /*
+      DifferentiationInvoker indirect(ai, this->original, attr);
```
Note: indirect differentiation invokers are currently not handled post-refactoring. I've disabled the test for now.

dan-zheng commented on lib/SILOptimizer/IPO/CapturePropagation.cpp:
```diff
@@ -461,13 +461,20 @@ bool CapturePropagation::optimizePartialApply(PartialApplyInst *PAI) {
-      // Notify the pass manager about the new function.
-      addFunctionToPassManagerWorklist(GenericSpecialized.first,
-                                       GenericSpecialized.second);
+    // SWIFT_ENABLE_TENSORFLOW
```
Yes, I also believe so.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -413,7 +432,6 @@ class PrimalInfo {
   }
 
 public:
-  PrimalInfo(const PrimalInfo &) = delete;
```
I deleted that line to unblock compilation errors I wasn't sure how to fix.

Here are the errors when I add the line back:
```
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:3545:19: error: call to deleted constructor of '(anonymous namespace)::PrimalInfo'
        vjp(vjp), primalInfo(primalInfo), invoker(invoker),
                  ^          ~~~~~~~~~~
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:432:3: note: 'PrimalInfo' has been explicitly marked deleted here
  PrimalInfo(const PrimalInfo &) = delete;
  ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5248:33: error: call to deleted constructor of '(anonymous namespace)::PrimalInfo'
                                *primalInfo.get(), invoker, activityInfo,
                                ^~~~~~~~~~~~~~~~~
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:432:3: note: 'PrimalInfo' has been explicitly marked deleted here
  PrimalInfo(const PrimalInfo &) = delete;
  ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:3540:38: note: passing argument to parameter 'primalInfo' here
                          PrimalInfo primalInfo,
                                     ^
```

The two copies happen in:
* The constructor of `AdjointEmitter`.
* The call site of the constructor of `AdjointEmitter` (`*primalInfo.get()`).

Do you have any suggestions?

rxwei replied:
When you take it as an argument, it should be `PrimalInfo &primalInfo` instead of `PrimalInfo primalInfo`.

dan-zheng replied:
That works, thanks!

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -739,12 +640,33 @@ class ADContext {
+
+  /// List of instructions referencing generated functions (VJPs and adjoints).
+  /// Saved for deletion during cleanup.
+  SmallVector<SILValue, 32> generatedFunctionReferences;
```
`generatedFunctionReferences` stores the `SILValue` outputs of `emitAssociatedFunctionReference`:
```
getGeneratedFunctionReferences().push_back(assocFn);
```

So `FunctionRefInst *` is not appropriate as the element type. I'll leave things as is.

dan-zheng replied:
The doc comment is not accurate, I'll try to improve it.

dan-zheng replied:
Improved doc comment and naming:
```
  /// List of associated function references, generated via
  /// `emitAssociatedFunctionReference`.
  /// Saved for deletion during cleanup.
  SmallVector<SILValue, 32> generatedAssociatedFunctionReferences;
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -870,6 +830,48 @@ class ADContext {
+  /// corresponding to the specified parameter indices. Returns nullptr if it
+  /// does not exist.
+  ///
+  /// TODO: Currently we are doing a O(n) lookup. This could be improved by
```
Good point, I'll remove the blurb.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -891,95 +893,44 @@ class ADContext {
+private:
+  /// Canonicalizes the given `autodiff_function` instruction, populating
+  /// missing associated functions, and returns
+  SILValue getCanonicalizedAutoDiffFunctionInst(
```
We discussed in person. The semantics of the method are a bit complicated.
We agreed to use the original name of this method, `promoteToDifferentiableFunction`, since the method always returns a `SILValue` with `@differentiable` function type.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -739,12 +640,33 @@ class ADContext {
-  SmallDenseMap<std::pair<SILFunction *, SILAutoDiffIndices>, unsigned>
-      enqueuedTaskIndices;
+  /// Queue of `autodiff_function` instructions.
+  SmallVector<AutoDiffFunctionInst *, 32> autoDiffFunctionInsts;
```
Sorry, I accidentally copied the old doc comment `Queue of differentiation tasks.` without editing.

`autoDiffFunctionInsts` is definitely a stack (processed via while loop and `pop_back`). I'll update the doc comment:
```cpp
  /// The worklist (stack) of `autodiff_function` instructions to be processed.
  SmallVector<AutoDiffFunctionInst *, 32> autoDiffFunctionInsts;
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5769,15 +6037,18 @@ bool ADContext::processAutoDiffFunctionInst(AutoDiffFunctionInst *adfi) {
-  adfi->replaceAllUsesWith(newADFI);
+  // Replace all uses of `adfi`.
+  adfi->replaceAllUsesWith(differentiableFnValue);
+  llvm::erase_if(getAutoDiffFunctionInsts(), [&](AutoDiffFunctionInst *i) {
```
We discussed in person. You had the idea of effectively setting `AutoDiffFunctionInst` to nullptr in-place, which I've implemented.
```
  // Delete all worklist occurrences of `adfi` by setting them to nullptr.
  // This is more efficient than APIs like `llvm::erase_if`.
  for (auto &inst : getAutoDiffFunctionInsts())
    if (inst == adfi)
      inst = nullptr;
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -344,7 +363,7 @@ struct DifferentiationInvoker {
   }
 
-  std::pair<ApplyInst *, DifferentiationTask *>
+  std::tuple<ApplyInst *, SILFunction *, SILDifferentiableAttr *>
```
The triple for indirect differentiation is ``(original `apply` instruction, original function, `[differentiable]` attribute)`` for the parent invoker - the original function and `[differentiable]` attribute can be used to index into `ADContext::invokers` to get the parent invoker.

Your comment made me realize storing the original function is redundant, because the original function can be simply retrieved from the `apply` instruction. Does storing a `std::pair<ApplyInst *, SILDifferentiableAttr *>` (for now) make sense to you?

rxwei replied:
FWIW, `SILDifferentiableAttr` actually stores the original function.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2980,26 +3030,97 @@ class PrimalGenCloner final
-    auto insertion = getDifferentiationTask()->getNestedApplyInfo().try_emplace(
-        ai, std::move(info));
-    assert(insertion.second && "Nested apply info already exists");
+    NestedApplyInfo info{indices, indices, /*originalPullbackType*/ None};
```
`desiredIndices` and `actualIndices` should be removed from `NestedApplyInfo`, now that `emitAssociatedFunctionReference` is no longer called within `visitApply`.

`actualIndices` are resolved in `ADContext::processAutoDiffFunctionInst`.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24845/commits/10522e5882b8c1e0714a9523827abc2a653a0187. Also removed original `SILFunction *` from `SILDifferentiableAttribute` `DifferentiationInvoker` case.

dan-zheng committed:
Address review comments and simplify.

dan-zheng commented on test/AutoDiff/autodiff_indirect_diagnostics.swift:
```diff
@@ -6,14 +6,13 @@
 
```
I'm actually not sure what's the expected behavior for these tests. I'll debug a bit to see what's going on.

At a glance, the changes seem unideal. The old diagnostics were more informative and appeared on the `gradient` user invocation.
```diff
--- a/test/AutoDiff/autodiff_indirect_diagnostics.swift
+++ b/test/AutoDiff/autodiff_indirect_diagnostics.swift
@@ -6,14 +6,13 @@

 // Test unmet generic requirements.

-// expected-error @+1 {{function is not differentiable}}
 @differentiable
-// expected-note @+1 {{when differentiating this function definition}}
 func generic<T: Differentiable & FloatingPoint>(_ x: T) -> T {
+  // expected-error @+2 {{expression is not differentiable}}
   // expected-note @+1 {{member is not differentiable because the corresponding protocol requirement is not '@differentiable'}}
   return x + 1
 }
-_ = gradient(at: 1.0, in: generic) // expected-error {{function is not differentiable}}
+_ = gradient(at: 1.0, in: generic)
```

dan-zheng replied:
I didn't quite make sense of the diagnostic behavior changes yet.
We plan to change/fix diagnostic behavior soon, so it would be good to revisit these changes as part of that follow-up.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5663,92 +5474,515 @@ class Differentiation : public SILModuleTransform {
+      LookUpConformanceInModule(swiftMod));
+    assert(tangentSpace && "No tangent space for this type");
+    switch (tangentSpace->getKind()) {
+      case VectorSpace::Kind::Vector: {
```
Please make case labels align with `switch`.

dan-zheng commented:
Ready for re-review. cc @rxwei 
I'm investigating why diagnostics changed in `test/AutoDiff/autodiff_indirect_diagnostics.swift`.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5663,92 +5474,515 @@ class Differentiation : public SILModuleTransform {
+  SmallVector<AllocStackInst *, 4> localAllocations;
+
+  // Build a `.zero` argument for the given `Differentiable`-conforming type.
+  auto buildZeroArgument = [&](SILType zeroSILType) {
```
This logic should be unified with existing logic used for `emitZeroIndirect` instead of being duplicated.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/24845/commits/646096e004e5525b1baefd17701bd4f07a55ad40.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5663,92 +5474,515 @@ class Differentiation : public SILModuleTransform {
-                                newArgs, ai->isNonThrowing());
-        return inst;
+        auto *newThunkRef = builder.createFunctionRef(loc, newThunk);
+        auto arguments = map<SmallVector<SILValue, 8>>(
```
```suggestion
        SmallVector<SILValue, 8> arguments(ai->getArguments().begin(),
                                           ai->getArguments().end())
```

dan-zheng replied:
Oh nice. I've been using the suboptimal `map<SmallVector<SILValue, ...>>` pattern in multiple places.

dan-zheng replied:
Done. Also changed another `map<SmallVector<...>>` call-site to `SmallVector` direct constructor call.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5819,44 +6061,31 @@ void Differentiation::run() {
+    context.getAutoDiffFunctionInsts().pop_back();
+    // Skip instructions that have been set to nullptr by
+    // `processAutoDiffFunctionInst`.
+    if (adfi)
```
Maybe an early return here?

```suggestion
    if (!adfi) continue;
```

dan-zheng replied:
I'm a fan of `if (!adfi) continue;` just for the uniform indentation. 😃

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -891,95 +875,44 @@ class ADContext {
+  /// Calls `getOrCreateLinearMapIndexSubsetThunk` to thunk the linear map
+  /// returned by the associated function.
+  std::pair<SILFunction *, SubstitutionMap>
+  getOrCreateAssociatedFunctionIndexSubsetThunk(
```
In SIL differentiation, "index subset" can mean both parameter indices and result indices. This name is a bit unclear to me.

```suggestion
  getOrCreateSubsetParametersThunkForAssociatedFunction(
```

dan-zheng replied:
Thanks! `IndexSubset` was definitely a suboptimal name.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -891,95 +875,44 @@ class ADContext {
+  /// Get or create an associated function index subset thunk from
+  /// `actualIndices` to `desiredIndices` for the given associated function
+  /// value and original function operand.
+  SILFunction *getOrCreateLinearMapIndexSubsetThunk(
```
```suggestion
  SILFunction *getOrCreateSubsetParametersThunkForLinearMap(
```

rxwei reviewed (approved):
The patch LGTM overall. Huge congrats on getting this done!

There are two necessary improvements I think should be done before merging:
1. The `AdditiveArithmetic.zero` emission logic in a thunk builder should be unified with what's already in `AdjointEmitter`.
2. I suspect there's a performance regression caused by extra `autodiff_function` and `autodiff_function_extract` instructions being generated, which could pose a barrier for later optimizations. Folding trivial `autodiff_function`-`autodiff_function_extract` can be trivially done near the logic that canonicalizes `autodiff_function` instructions in the main loop.

Other than those, we should talk about fixing diagnostics for indirect calls soon.

dan-zheng committed:
Address review comments.

- Create common function `emitZeroIntoBuffer`.
  - Shared by `AdjointEmitter::emitZeroIndirect` and `buildZeroArgument` lambda.
- Minor naming/style changes.

dan-zheng committed:
Add `autodiff_function_extract` folding optimization.

Fold `autodiff_function_extract` users of `autodiff_function` instructions,
directly replacing them with operands of the `autodiff_function` instruction.

If the `autodiff_function` instruction has no non-`autodiff_function_extract`
users, delete the instruction itself after folding.

The `differentiation-skip-folding-autodiff-function-extraction` flag disables
folding for SIL testing purposes.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5505,149 +5333,132 @@ void DifferentiationTask::createJVP(bool isExported) {
-                          original->getLocation(), original->isBare(),
-                          IsNotTransparent, original->isSerialized(),
-                          original->isDynamicallyReplaceable());
+  auto linkage = autodiff::getAutoDiffFunctionLinkage(
```
Done.

dan-zheng commented:
Thanks @rxwei for the thorough review!

1. Unify `zero` emission logic is done in https://github.com/apple/swift/pull/24845/commits/646096e004e5525b1baefd17701bd4f07a55ad40.
2. `autodiff_function_extract` folding optimization is done in https://github.com/apple/swift/pull/24845/commits/064e61e6bf1c76b5d1813c8cec552bb0bd5b9f6e.
    * As we discussed, a `-Xllvm -differentiation-skip-folding-autodiff-function-extraction` flag was added to disable folding for SIL testing purposes. Used in three tests.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
